### ISSUE TITLE
Auto formats small default floats values into scientific notation breaking prisma syntax

### ISSUE THREAD

**Issue by sitch** [zenstackhq/zenstack#646]
**Description and expected behavior**

Very small floats `< 0.000001` and smaller translate into scientific notation in the `@default` field, reguardless of `Decimal` or `Float` 

*Zenstack*

```zenstack
model Example {
  epsilon Decimal @default(0.00000001)
}
```

Becomes:

*Prisma*

```prisma
model Example {
  epsilon Decimal @default(1e-8)
}
```

Which is invalid because prisma doesn't support scientific notation 

**Environment (please complete the following information):**

-   ZenStack version: [latest beta 20]
-   Prisma version: [e.g., 4.8
-   Database type: [e.g. Postgresql]

**Comment by ymc9** [comment]
Thanks, good catch @sitch ! Will include it in the next release. Let me know if you're interested in making a PR 😄.

**Comment by sitch** [comment]
So after a little more investigation, this is fundamentally a JS quirk (what a lovely language).  Any float `1e-7` or smaller automatically gets translated into scientific notation.

```js
num = 0.000001
console.log(num)
// 0.000001
```

```js
num = 0.0000001
console.log(num)
// 1e-7
```

There are some hacks that would work most of the time, but aren't concretely correct due to precision that has already been lost at parse time

```ts
function inferredPrecisionFloatToString(num: Number) : String {
  const exp = parseInt(num.toString().split("e-").slice(-1) ?? "0")
  return exp > 0 ? num.toFixed(exp) : num.toString()
}
```
So you would have:
```ts
inferredPrecisionFloatToString(0.00000001)
// "0.00000001"
```

Seems like this function could just be applied here: https://github.com/zenstackhq/zenstack/blob/02a4a179b8c32d0320b27adbefca196cd983dd05/packages/schema/src/plugins/prisma/prisma-builder.ts#L293

But a better solution would be to get the raw token from the AST parse and use that or to not actually interpret the token as Number, but with Decimal internally

**Comment by sitch** [comment]
Also, a small discrepancy but:

In `prisma` you can do:
```prisma
model Example {
  id String @id
  ep1 Decimal @default("0.00000001")
  ep2 Decimal @default(0.00000001)
  ep3 Float @default(0.00000001)
}
```
But in `zmodel`:
```prisma
model Example {
  id String @id
  ep1 Decimal @default("0.00000001") // doesn't work `Value is not assignable to parameter`
  ep2 Decimal @default(0.00000001)
  ep3 Float @default(0.00000001)
}
```

**Comment by sitch** [comment]
Basically a "correct" solution becomes hairy at the prisma level bc a column can have custom precision settings which would then require validating the attribute value to see if it's string representation is representable in float space, given the columns precision settings

**Comment by ymc9** [comment]
Many thanks for putting time into investigating this @sitch ! I agree with you the best solution seems to use the raw token to output to prisma schema, so at least we can be completely consistent with Prisma ... Yes, passing string values also need to be fixed.

Now the issue sounds more complex than I initially thought. I can further check how to change the parser part, but let me know if you plan to dig deeper there so we don't spend duplicated efforts 😄.

**Comment by ymc9** [comment]
Btw is this blocking you right now? A workaround could be:

```prisma
model Example {
  epsilon Decimal @prisma.passthrough('@default(0.00000001)')
}
```

**Comment by sitch** [comment]
> Btw is this blocking you right now? A workaround could be:
> 
> ```prisma
> model Example {
>   epsilon Decimal @prisma.passthrough('@default(0.00000001)')
> }
> ```

This is what I'm currently doing

**Comment by sitch** [comment]
I think just updating the parser should be good. Also the zenstack number functions currently apply to decimal as well, but I'll open a separate ticket for that! Cheers @ymc9

**Comment by ymc9** [comment]
Fixed in [1.0.0-beta.21](https://github.com/zenstackhq/zenstack/releases/tag/v1.0.0-beta.21)
